File: wallendar-app/app/src/main/java/com/td/wallendar/addmembers/ui/AddMembersPresenter.java
package com.td.wallendar.addmembers.ui;

import com.td.wallendar.dtos.request.AddMembersRequest;
import com.td.wallendar.models.Group;
import com.td.wallendar.repositories.interfaces.GroupsRepository;
import com.td.wallendar.utils.scheduler.AndroidSchedulerProvider;
import com.td.wallendar.utils.scheduler.SchedulerProvider;

import java.lang.ref.WeakReference;
import java.util.HashSet;
import java.util.List;

import io.reactivex.disposables.CompositeDisposable;

public class AddMembersPresenter {

    private final WeakReference<AddMembersView> view;
    private final GroupsRepository groupsRepository;
    private final CompositeDisposable disposable;
    private final SchedulerProvider schedulerProvider;

    public AddMembersPresenter(final AddMembersView addMembersView, final GroupsRepository groupsRepository) {
        view = new WeakReference<>(addMembersView);
        this.groupsRepository = groupsRepository;
        this.disposable = new CompositeDisposable();
        this.schedulerProvider = new AndroidSchedulerProvider();
    }

    public void submitMembers(final long groupId, final List<String> emailMembers) {
        disposable.add(groupsRepository.addMembers(groupId, new AddMembersRequest(new HashSet<>(emailMembers)))
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::onMembersAddedSuccessfully, this::onMembersAddedWithError));
    }

    private void onMembersAddedWithError(Throwable throwable) {
        view.get().onMembersAddedWithError();
    }

    private void onMembersAddedSuccessfully(Group group) {
        view.get().onMembersAddedSuccessfully(group.getId());
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/addmembers/ui/AddMembersPresenter.kt
package com.td.wallendar.addmembers.ui

import com.td.wallendar.dtos.request.AddMembersRequest
import com.td.wallendar.repositories.interfaces.EventsRepository
import com.td.wallendar.repositories.interfaces.GroupsRepository
import com.td.wallendar.utils.scheduler.SchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference

class AddMembersPresenter(
        addMembersView: AddMembersView,
        private val groupsRepository: GroupsRepository,
        private val eventsRepository: EventsRepository,
        private val schedulerProvider: SchedulerProvider,
        private val isEvent: Boolean,
) {
    private val view: WeakReference<AddMembersView> = WeakReference(addMembersView)
    private val disposable: CompositeDisposable = CompositeDisposable()
    fun submitMembers(groupId: Long, emailMembers: MutableList<String>) {
        val repositoryCall = {
            if (isEvent) {
                eventsRepository.addMembers(groupId, AddMembersRequest(HashSet(emailMembers)))
            } else {
                groupsRepository.addMembers(groupId, AddMembersRequest(HashSet(emailMembers)))
            }
        }
        disposable.add(repositoryCall()
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe({ onMembersAddedSuccessfully() }) { throwable: Throwable -> onMembersAddedWithError(throwable) })
    }

    fun onViewDetached() {
        disposable.dispose()
    }

    private fun onMembersAddedWithError(throwable: Throwable) {
        view.get()?.onMembersAddedWithError()
    }

    private fun onMembersAddedSuccessfully() {
        view.get()?.onMembersAddedSuccessfully()
    }

}